File: test/unit/storage.test.js
'use strict';
const { describe, it, after } = require('node:test');
const assert = require('node:assert/strict');
const fs = require('fs');
const os = require('os');
const path = require('path');
const { createLocalStorage } = require('../../src/storage');

function makeTmpDir() {
    return path.join(os.tmpdir(), 'test-storage-' + Date.now() + '-' + Math.random().toString(36).slice(2));
}

function writeStream(stream, content) {
    return new Promise((resolve, reject) => {
        stream.on('finish', resolve);
        stream.on('error', reject);
        stream.write(content);
        stream.end();
    });
}

function readStream(stream) {
    return new Promise((resolve, reject) => {
        let data = '';
        stream.on('data', chunk => { data += chunk; });
        stream.on('end', () => resolve(data));
        stream.on('error', reject);
    });
}

describe('createLocalStorage', () => {
    const dirs = [];

    after(() => {
        for (const dir of dirs) {
            fs.rmSync(dir, { recursive: true, force: true });
        }
    });

    it('write then read round-trip', async () => {
        const dir = makeTmpDir();
        dirs.push(dir);
        const s = createLocalStorage(dir);
        await writeStream(s.file('artifact.jar').createWriteStream(), 'hello world');
        const content = await readStream(s.file('artifact.jar').createReadStream());
        assert.equal(content, 'hello world');
    });

    it('exists() returns [true] for existing file', async () => {
        const dir = makeTmpDir();
        dirs.push(dir);
        const s = createLocalStorage(dir);
        await writeStream(s.file('present.txt').createWriteStream(), 'data');
        const [exists] = await s.file('present.txt').exists();
        assert.equal(exists, true);
    });

    it('exists() returns [false] for missing file', async () => {
        const dir = makeTmpDir();
        dirs.push(dir);
        const s = createLocalStorage(dir);
        const [exists] = await s.file('missing.txt').exists();
        assert.equal(exists, false);
    });

    it('getMetadata() returns timeCreated as valid ISO string', async () => {
        const dir = makeTmpDir();
        dirs.push(dir);
        const s = createLocalStorage(dir);
        await writeStream(s.file('meta.txt').createWriteStream(), 'data');
        const [meta] = await s.file('meta.txt').getMetadata();
        assert.equal(typeof meta.timeCreated, 'string');
        assert.ok(!isNaN(new Date(meta.timeCreated).getTime()), 'timeCreated should be a valid date');
    });

    it('delete() removes the file', async () => {
        const dir = makeTmpDir();
        dirs.push(dir);
        const s = createLocalStorage(dir);
        await writeStream(s.file('to-delete.txt').createWriteStream(), 'bye');
        await s.file('to-delete.txt').delete();
        const [exists] = await s.file('to-delete.txt').exists();
        assert.equal(exists, false);
    });

    it('createWriteStream creates nested directories', async () => {
        const dir = makeTmpDir();
        dirs.push(dir);
        const s = createLocalStorage(dir);
        await writeStream(s.file('deep/nested/path/file.txt').createWriteStream(), 'nested');
        const [exists] = await s.file('deep/nested/path/file.txt').exists();
        assert.equal(exists, true);
    });

    it('createReadStream on missing file emits error event', async () => {
        const dir = makeTmpDir();
        dirs.push(dir);
        const s = createLocalStorage(dir);
        const stream = s.file('nonexistent.txt').createReadStream();
        await new Promise((resolve, reject) => {
            stream.on('error', () => resolve());
            stream.on('end', () => reject(new Error('Expected error but got end')));
            stream.resume(); // trigger read
        });
    });
});

File: storage_test.go
package main

import (
	"context"
	"io"
	"strings"
	"testing"
	"time"
)

func TestStorage(t *testing.T) {
	ctx := context.Background()

	t.Run("write then read round-trip", func(t *testing.T) {
		dir := t.TempDir()
		s := NewLocalStorage(dir)
		w := s.File("artifact.jar").NewWriter(ctx)
		if _, err := io.Copy(w, strings.NewReader("hello world")); err != nil {
			t.Fatal(err)
		}
		if err := w.Close(); err != nil {
			t.Fatal(err)
		}
		r, err := s.File("artifact.jar").NewReader(ctx)
		if err != nil {
			t.Fatal(err)
		}
		defer r.Close()
		data, err := io.ReadAll(r)
		if err != nil {
			t.Fatal(err)
		}
		if string(data) != "hello world" {
			t.Errorf("got %q, want %q", data, "hello world")
		}
	})

	t.Run("Exists returns true for existing file", func(t *testing.T) {
		dir := t.TempDir()
		s := NewLocalStorage(dir)
		w := s.File("present.txt").NewWriter(ctx)
		_, _ = w.Write([]byte("x"))
		_ = w.Close()
		ok, err := s.File("present.txt").Exists(ctx)
		if err != nil {
			t.Fatal(err)
		}
		if !ok {
			t.Error("expected Exists=true for written file")
		}
	})

	t.Run("Exists returns false for missing file", func(t *testing.T) {
		dir := t.TempDir()
		s := NewLocalStorage(dir)
		ok, err := s.File("nonexistent.txt").Exists(ctx)
		if err != nil {
			t.Fatal(err)
		}
		if ok {
			t.Error("expected Exists=false for missing file")
		}
	})

	t.Run("GetMetadata returns valid TimeCreated", func(t *testing.T) {
		dir := t.TempDir()
		s := NewLocalStorage(dir)
		before := time.Now().Add(-time.Second)
		w := s.File("meta.txt").NewWriter(ctx)
		_, _ = w.Write([]byte("x"))
		_ = w.Close()
		meta, err := s.File("meta.txt").GetMetadata(ctx)
		if err != nil {
			t.Fatal(err)
		}
		if meta.TimeCreated.IsZero() {
			t.Error("TimeCreated is zero")
		}
		if meta.TimeCreated.Before(before) {
			t.Errorf("TimeCreated %v is before test start %v", meta.TimeCreated, before)
		}
	})

	t.Run("Delete removes file", func(t *testing.T) {
		dir := t.TempDir()
		s := NewLocalStorage(dir)
		w := s.File("delete-me.txt").NewWriter(ctx)
		_, _ = w.Write([]byte("x"))
		_ = w.Close()
		if err := s.File("delete-me.txt").Delete(ctx); err != nil {
			t.Fatal(err)
		}
		ok, err := s.File("delete-me.txt").Exists(ctx)
		if err != nil {
			t.Fatal(err)
		}
		if ok {
			t.Error("expected Exists=false after delete")
		}
	})

	t.Run("NewWriter creates nested directories", func(t *testing.T) {
		dir := t.TempDir()
		s := NewLocalStorage(dir)
		w := s.File("deep/nested/path/file.txt").NewWriter(ctx)
		_, _ = w.Write([]byte("nested"))
		_ = w.Close()
		ok, err := s.File("deep/nested/path/file.txt").Exists(ctx)
		if err != nil {
			t.Fatal(err)
		}
		if !ok {
			t.Error("expected Exists=true for nested file")
		}
	})

	t.Run("NewReader on missing file returns error", func(t *testing.T) {
		dir := t.TempDir()
		s := NewLocalStorage(dir)
		_, err := s.File("missing.txt").NewReader(ctx)
		if err == nil {
			t.Error("expected error for missing file, got nil")
		}
	})
}
